File: 789-kth-largest-element-in-a-stream/kth-largest-element-in-a-stream.py
class KthLargest:

    def __init__(self, k: int, nums: List[int]):
        self.nums = nums
        self.k = k

        heapq.heapify(self.nums) # [2,4,5,8]  

    def add(self, val: int) -> int:
        heapq.heappush(self.nums,val) # [2,3,4,5,8]

        while self.nums and len(self.nums) != self.k:
            heapq.heappop(self.nums)
        
        # [2,3,4]

        return self.nums[0]
        
        # SC -> O(N)
        # TC -> O(N Log K)
        
        


# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: 789-kth-largest-element-in-a-stream/kth-largest-element-in-a-stream.java
class KthLargest {
    int k;
    PriorityQueue<Integer> pq;

    public KthLargest(int k, int[] nums) {
        this.k = k;
        this.pq = new PriorityQueue<>((a, b)-> a - b);
        for (int num: nums) this.pq.offer(num);
        System.out.println(this.pq);
    }
    
    public int add(int val) {
        this.pq.offer(val);
        while(this.pq.size() > this.k){
            this.pq.poll();
        }

        return this.pq.peek();
    }
}

// SC -> O(N)
// TC -> O(N)

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest obj = new KthLargest(k, nums);
 * int param_1 = obj.add(val);
 */